karate-core/src/main/java/com/intuit/karate/driver/DriverOptions.java

```diff
@@ -603,6 +603,9 @@ public String scriptAllSelector(String locator, String expression, String contex
         boolean isXpath = locator.startsWith("/");
         String selector;
         if (isXpath) { // XPathResult.ORDERED_NODE_ITERATOR_TYPE = 5
+            if (!DOCUMENT.equals(contextNode)) {
```

Comment:
I'm not sure this is going to work. shouldn't line 603 be changed to include startsWith `./` and `../`

and we definitely need tests. I've made it easier for you with this commit: https://github.com/karatelabs/karate/commit/5b4505d97f343ca72e8c77a8d2a8237d03028eb5

please can you add a few examples and edge cases. and you can run `LocalSingleRunner.java` to test locally. make sure existing XPath is not broken of course. and then once tests pass in CI / DC we can merge

Reply:
Yes, my first solution for Issue#1941 was a method like:

```
private static boolean isXPath(String locator) {
        return locator.matches( "^[.]{0,2}/.+$");
}
```
... but, the current implementation in `DriverOptions.java` do not allow this solution:

```
 public static String selector(String locator, String contextNode) {
[...]
            } else if (!DOCUMENT.equals(contextNode)) {
                locator = "." + locator; // evaluate relative to this node not root
            }
[...]
    }
```

**The source code "detects" when the context does not match the document root and then sets the path relative.**

This additional code was missing in:
`public String scriptAllSelector(String locator, String expression, String contextNode)`

With this bugfix I have added the missing part.

Reply:
For my project it's not necessary to support "../"

I just want to enable the relative XPath locator with "./" for "singleSelector" and "allSelector". Even if the the point isn't a visible part of the parameter "locator".

Reply:
I'm simply not able to understand this, so I'll wait until I see some tests that demonstrate what you mean

Reply:
My task is to find document nodes by XPath
- absolute and
- relative

Karate provides two method categories for this:
- locate, script => **selector**
- locateAll, scriptAll => **scriptAllSelector**

Besides the parameter "expression", there are the parameter
- **locator** => detects XPath expression, if it starts with "/".
- **contextNode** => optional, with fallback document root

According to XPath:
- an absolute path e.g.: `/html/body/header` or `//header`
- whereas a relative path e.g.: `./h1`

With Karate I'm able to select the node "header":
`def headerNode = locate( "/html/body/header") `

..., if I want to search for a contained "h1" starting from the selection "headerNode", I have to use the parameter "expression":
`def h1Node = headerNode.locate( "./h1")` => **wrong** relative expression with karate

The implementation of the method "selector" (DriverOptions.java) detects XPath expressions only by a leading "/" character. Therefore, I initially assumed that relative path specifications are not provided (see issue #1941). However, the method contains a internal logic which automatically prefixes the missing character "." depending on the current "contextNode".

So the correct search (with karate) is without the leading dot:
`def h1Node = headerNode.locate( "/h1")` => **correct** relative expression with karate

**Previously, context detection was only included in the method "selector" and was missing in "scriptAllSelector".**

===

I know the relative XPath expression without the leading dot is a bit confusing, but on the other hand the implemented detection has its benefit. A documentation would certainly be important at this point.

Reply:
> missing in "scriptAllSelector".

so please write a test that shows this failing

> A documentation would certainly be important at this point.

excellent. please include a change to the docs in your PR

Reply:
> so please write a test that shows this failing

please have a look at [karate-e2e-tests/src/test/java/driver/16.feature]

Reply:
sorry I'm still lost, maybe it is an issue on my side.
is there a code change needed to karate at all or just documentation ?

Reply:
> is there a code change needed to karate at all or just documentation ?

Both, 
(a) the bugfix is a part of this PR and 
(b) the handling with relative XPath expressions have to be documented more clearly

Reply:
@BG04129 ok give me some time now, and I'll do it. thanks for providing the details. I really want the tests to be clear with comments (and in the same file 15.feature) so let me take care of that. I might do the code changes myself and close this PR, so heads up.